runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java

```diff
@@ -207,6 +207,58 @@ public void removeProcessingTimer(KeyedTimerData<K> keyedTimerData) {
     state.deletePersisted(keyedTimerData);
   }
 
+  /** Functional interface for firing a timer. */
+  @FunctionalInterface
+  public interface TimeFiringFn<K> {
+    void fire(KeyedTimerData<K> timerData);
+  }
+
+  /**
+   * Consolidated method to fire ready timers. It loops while the event time buffer has timers that
+   * are due based on the current input watermark. For each timer collected, it checks if the timer
+   * still exists in state; if so, it deletes it and fires it. Finally, if the event time buffer is
+   * empty, it reloads timers from state.
+   */
+  public void fireReadyTimers(TimeFiringFn<K> firingFn) {
+    // Process timers as long as there are timers in the buffer that are due.
+    while (!eventTimeBuffer.isEmpty()
+        && !eventTimeBuffer.first().getTimerData().getTimestamp().isAfter(inputWatermark)) {
+
+      // Create a list of ready timers without deleting them from state yet.
+      Collection<KeyedTimerData<K>> readyTimers = new ArrayList<>();
+      while (!eventTimeBuffer.isEmpty()
+          && !eventTimeBuffer.first().getTimerData().getTimestamp().isAfter(inputWatermark)
+          && readyTimers.size() < maxReadyTimersToProcessOnce) {
+        readyTimers.add(eventTimeBuffer.pollFirst());
+      }
+
+      // For each timer in our ready list, verify it still exists in state.
+      for (KeyedTimerData<K> keyedTimerData : readyTimers) {
+        Long storedTimestamp = state.get(keyedTimerData);
+        if (storedTimestamp != null
+            && storedTimestamp.equals(keyedTimerData.getTimerData().getTimestamp().getMillis())) {
+          // The timer still exists in state; delete it now and fire it.
+          state.deletePersisted(keyedTimerData);
+          firingFn.fire(keyedTimerData);
+        }
+      }
+
+      // Optionally, warn if we hit the processing limit while there are still expired timers.
+      if (readyTimers.size() == maxReadyTimersToProcessOnce
+          && !eventTimeBuffer.isEmpty()
+          && eventTimeBuffer.first().getTimerData().getTimestamp().isBefore(inputWatermark)) {
+        LOG.warn(
+            "Loaded {} expired timers, the remaining will be processed at next watermark.",
+            maxReadyTimersToProcessOnce);
+      }
```

Comment:
I think we need a "break" statement here. Otherwise it's going to do the loop again and fetch the next batch of readyTimers. What do you think?